plugins/radar-android-google-sleep/src/main/java/org.radarbase.passive.google.sleep/GoogleSleepManager.kt

```diff
@@ -0,0 +1,153 @@
+        sleepBroadcastReceiver = SleepReceiver(this)
+    }
+
+    override fun start(acceptableIds: Set<String>) {
+        register()
+        sleepHandler.start()
+        status = SourceStatusListener.Status.READY
+        sleepHandler.execute {
+            createSleepPendingIntent()
+            registerForSleepData()
+        }
+    }
+
+     fun sendSleepClassifyData(sleepIntent: Intent) {
+        logger.info("Sleep segment event data received")
+        val sleepSegmentEvents: List<SleepSegmentEvent> = SleepSegmentEvent.extractEvents(sleepIntent)
+        sleepSegmentEvents.forEach { sleepSegmentEvent ->
+            val time = currentTime
+            val sleepStartTime: Double = sleepSegmentEvent.startTimeMillis / 1000.0
+            val sleepEndTime: Double = sleepSegmentEvent.endTimeMillis / 1000.0
+            val sleepDuration: Double = sleepSegmentEvent.segmentDurationMillis / 1000.0
+            val sleepStatus: SleepStatus = sleepSegmentEvent.status.toSleepStatus()
+
+            send(segmentEventTopic, GoogleSleepSegmentEvent(time, time, sleepStartTime,
+            sleepEndTime, sleepDuration, sleepStatus))
+        }
+    }
+
+     fun sendSleepSegmentData(sleepIntent: Intent) {
+        logger.info("Sleep classify event data received")
+        val sleepClassifyEvents: List<SleepClassifyEvent> = SleepClassifyEvent.extractEvents(sleepIntent)
+        sleepClassifyEvents.forEach {  sleepClassifyEvent ->
+            val time = sleepClassifyEvent.timestampMillis / 1000.0
+            val sleepConfidence = sleepClassifyEvent.confidence
+            val light = sleepClassifyEvent.light
+            val motion = sleepClassifyEvent.motion
+
+            send(classifyEventTopic, GoogleSleepClassifyEvent(time, currentTime, sleepConfidence, light, motion))
+        }
+    }
+
+    @SuppressLint("MissingPermission")
+    private fun registerForSleepData() {
+        if ( isPermissionGranted() ) {
+            status = SourceStatusListener.Status.CONNECTING
+
+            ActivityRecognition.getClient(service)
+                .requestSleepSegmentUpdates(
+                    sleepPendingIntent,
+                    SleepSegmentRequest.getDefaultSleepSegmentRequest() )
+                .addOnSuccessListener {
+                    status = SourceStatusListener.Status.CONNECTED
+                    logger.info("Successfully subscribed to sleep data")
+                }
+                .addOnFailureListener {exception ->
+                    status = SourceStatusListener.Status.UNAVAILABLE
+                    logger.error("Exception while subscribing to sleep data: $exception")
+                }
+        }
+        else logger.warn("Permission not granted for ACTIVITY_RECOGNITION")
+    }
+
+    private fun isPermissionGranted(): Boolean {
+        val permission = if ( Build.VERSION.SDK_INT >= 29 ) Manifest.permission.ACTIVITY_RECOGNITION
+        else "com.google.android.gms.permission.ACTIVITY_RECOGNITION"
```

Comment:
In other cases I made this a `val` in the companion object
```kotlin
val ACTIVITY_RECOGNITION_COMPAT = if (Build.VERSION.SDK_INT >= 29) Manifest.permission.ACTIVITY_RECOGNITION
        else "com.google.android.gms.permission.ACTIVITY_RECOGNITION"
```
Perhaps define that in the companion object of `GoogleSleepProvider` and reuse here.